File: 49.Group_Anagrams/49.Group_Anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for(var str: strs){
            char[] charArrayKey = str.toCharArray();
            Arrays.sort(charArrayKey);
            String key = new String(charArrayKey);
            //取出key对应的stringList，如果没有则创建一个新的
            List<String> stringList = map.getOrDefault(key, new ArrayList<String>());
            //在stringList中添加新的string以后重新放回哈希表中
            stringList.add(str);
            map.put(key, stringList);
        }
        
        return new ArrayList<List<String>>(map.values());
    }
}

File: 49.Group_Anagrams/49.Group_Anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> hash_table;
        for(auto str: strs){
            string key = str;
            sort(key.begin(), key.end());
            hash_table[key].push_back(str);
        }
        vector<vector<string>> result;
        for(auto element: hash_table){
            result.push_back(element.second);
        }
        return result;
    }
};


